otj-metrics-core/src/main/java/com/opentable/metrics/OTInstrumentedHandler.java

```diff
@@ -0,0 +1,343 @@
+            }
+
+            @Override
+            public void onStartAsync(AsyncEvent event) throws IOException {
+                event.getAsyncContext().addListener(this);
+            }
+
+            @Override
+            public void onError(AsyncEvent event) throws IOException {
+            }
+
+            @Override
+            public void onComplete(AsyncEvent event) throws IOException {
+                final HttpChannelState state = ((AsyncContextEvent)event).getHttpChannelState();
+
+                final Request request = state.getBaseRequest();
+                final long elapsed = System.currentTimeMillis() - request.getTimeStamp();
+
+                updateResponses(request, elapsed);
+
+                activeSuspended.dec();
+            }
+
+        };
+    }
+
+    /**
+     * Jetty processes new requests and resumed (previously suspended) requests by the following logic:
+     *
+     * - New requests (isInitial() == true) are dispatched to the handler.
+     * - Previously suspended requests (isInitial() == false) are redispatched to the handler upon being resumed.
+     *
+     * The handle method receives a dispatched request in one of these two states, attempts to handle it, and then
+     * finally updates the appropriate metrics.
+     *
+     * After the handle method is called, the request is left in one of three states:
+     *
+     * - Requests that got suspended during handling (isSuspended() == true) get attached with an AsyncListener,
+     *   do not update response metrics, and will later be redispatched to the handler whenever processing is resumed.
+     * - New requests that completed without being suspended and which were not previously suspended
+     *   (isSuspended() == false && isInitial() == true) will trigger the update of response metrics in-line.
+     * - Previously suspended requests that finally got completely handled (isSuspended() == true && isInitial() == false)
+     *   will be picked up by the attached AsyncListener's onComplete method, and trigger the update of response metrics.
+     */
+    @Override
+    public void handle(String path,
+                       Request request,
+                       HttpServletRequest httpRequest,
+                       HttpServletResponse httpResponse) throws IOException, ServletException {
+
+        activeDispatches.inc();
+
+        final long start;
+        final HttpChannelState state = request.getHttpChannelState();
+        if (state.isInitial()) {
+            // New request
+            activeRequests.inc();
+            start = request.getTimeStamp();
+        }
+        else {
+            // Resumed request
+            start = System.currentTimeMillis();
+            asyncDispatches.mark();
+        }
+
+        try {
+            super.handle(path, request, httpRequest, httpResponse);
+        }
+        finally {
+            final long now = System.currentTimeMillis();
+            final long dispatched = now - start;
```

Comment:
System.currentTimeMillis() is not monotonic. This could cause issues with things like leap seconds and other horrible things.
I suggest you use one of the following:

- https://google.github.io/guava/releases/23.0/api/docs/com/google/common/base/Stopwatch.html
- https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/time/StopWatch.html
- `System.nanotime()`

Reply:
This was taken directly from DropWizard Metrics. Let me post a diff to show what was changed.

Reply:
As discussed offline, we will continue to use `System.currentTimeMillis()` for the following reasons:

We can't use `Stopwatch` because we have to compare durations where the start time is obtained outside of the current scope (i.e. from Jetty `Request#getTimestamp`, and inside an AsyncListener that might be dispatched, suspended, and resumed multiple times.

We can't use `System.nanotime` because you can only compare a nanoTime to another nanoTime, which indicates a duration. But the object we need to compare this to is a Jetty`Request#getTimestamp` that represents wall clock time in milliseconds in UTC.